### ISSUE TITLE
else statement returns still emits a return a value error

### ISSUE THREAD

**Issue by barkthins** [xjamundx/eslint-plugin-promise#8]
The error `each then() should return a value` is reported even though the `.then` always returns a promise in the code below.

Apparently the logic for following return values can't evaluate an else clause
        
        return oracle.getPrices(c)
        .then(function(p) {
            if (p.price != 0) {
                return p.price.toNumber() / 1e6;
            } else {
                // XXX unavoidable race
                return 200;
            }
        });

**Comment by xjamundx** [comment]
right because I personally prefer the `no-else-return` rule, so in my case I'd still do the `return 200` at the root level of the function.

Yeah totally valid bug though, solution is tricky to enforce properly.

We could check that there's *any* `return` inside of a `then` function. For example:

```
.then(function(p) {
    if (p < 10) {
        return p * 1000;
    }
})
```

But then it obviously misses cases like this where the root of the function isn't covered, but to some extent that is covered by `consistent-return`.

What do you think? Would you be able to try to make the code change?

**Comment by barkthins** [comment]
funny thing is worked on medical devices and prefer `else-return`.  Please cite why `no-else-return` is a rule...

If you don't have enough of the parse tree to figure out all the cases possible and make sure there's a return for each case, then you probably can't solve the problem.  I don't know enough about how eslint works to know whether that's possible.

**Comment by xjamundx** [comment]
@barkthins I use `no-else-return` in javascript, because of the endless callback functions. This saves one level of nesting on the `else` path

As far as creating custom ESLint rules, we do have a full tree, it's just the tools for traversing it are pretty weak, so checking every path is pretty difficult. [Code path analysis](http://eslint.org/docs/developer-guide/code-path-analysis/) landed which is supposed to make this easier and may be worth checking out.

I'd love your help adding this ability to the rule, let me know if you can do it!

**Comment by barkthins** [comment]
@xjamundx I use promises. What's this issue with callback nesting you speak of? ;-P

My plate is stacked high and deep, being at a startup.   I'm busy enough answering Ethereum questions, I'm afraid another community will give me overload...

**Comment by xjamundx** [comment]
K I'll keep this open until I get some time to figure it out!

**Comment by scottnonnenberg** [comment]
Not exactly the same, but very similar - I get an 'each then() should return a value' error for this:

```javascript
.then(([response, text]) => {
  try {
    return [response, parseWithDates(text)];
  }
  catch (e) {
    return [response];
  }
})
```

As well as for this, from my tests:

```javascript
return fetch.default()
  .then(() => {
    throw new Error('expected error to be thrown');
  })
  .catch(error => {
    expect(error).to.have.property('message', 'Something went wrong!');
  });
````

**Comment by xjamundx** [comment]
ahh thanks for the note @scottnonnenberg I'll try to look at this..in the meantime you may have to do this:
```js
var dates
try { dates = parseWithDates(text) }
catch (e) { data = undefined }
return [response, dates
```
